### ISSUE TITLE
Running error when using HuggingFaceModelWrapper and PWWS attacker

### ISSUE THREAD

**Issue by zjiehang** [QData/TextAttack#249]
I met error when runing the PWWS Attack on huggingface model. My code is showed below: 

```python
custom_dataset = [
    ('Malaria deaths in Africa fall by 5% from last year', 0),
    ('Washington Nationals defeat the Houston Astros to win the World Series', 1),
    ('Exxon Mobil hires a new CEO', 0),
    ('Microsoft invests $1 billion in OpenAI', 1),
]
model = AutoModelForSequenceClassification.from_pretrained(model_path)
tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)

model_wrapper = HuggingFaceModelWrapper(model, tokenizer, batch_size=100)
attacker = PWWSRen2019.build(model_wrapper)
results_iterable = attacker.attack_dataset(custom_dataset)

logger = CSVLogger(color_method='html')
num_successes = 0
for result in results_iterable:
    if isinstance(result, SuccessfulAttackResult):
        logger.log_attack_result(result)
        num_successes += 1
        print(f'{num_successes} of 10 successes complete.')
```

And the error is:
```
Traceback (most recent call last):
  File "/root/CLUE/utils/textattack_test.py", line 36, in <module>
    for result in results_iterable:
  File "/root/miniconda/lib/python3.7/site-packages/textattack/shared/attack.py", line 266, in attack_dataset
    for goal_function_result in examples:
  File "/root/miniconda/lib/python3.7/site-packages/textattack/shared/attack.py", line 251, in _get_examples_from_dataset
    attacked_text, ground_truth_output
  File "/root/miniconda/lib/python3.7/site-packages/textattack/goal_functions/goal_function.py", line 55, in init_attack_example
    result, _ = self.get_result(attacked_text, check_skip=True)
  File "/root/miniconda/lib/python3.7/site-packages/textattack/goal_functions/goal_function.py", line 66, in get_result
    results, search_over = self.get_results([attacked_text], **kwargs)
  File "/root/miniconda/lib/python3.7/site-packages/textattack/goal_functions/goal_function.py", line 83, in get_results
    model_outputs = self._call_model(attacked_text_list)
  File "/root/miniconda/lib/python3.7/site-packages/textattack/goal_functions/goal_function.py", line 180, in _call_model
    outputs = self._call_model_uncached(uncached_list)
  File "/root/miniconda/lib/python3.7/site-packages/textattack/goal_functions/goal_function.py", line 149, in _call_model_uncached
    outputs = self.model(inputs)
  File "/root/miniconda/lib/python3.7/site-packages/textattack/models/wrappers/huggingface_model_wrapper.py", line 44, in __call__
    model_predict, ids, batch_size=self.batch_size
  File "/root/miniconda/lib/python3.7/site-packages/textattack/shared/utils/tensor.py", line 14, in batch_model_predict
    batch_preds = model_predict(batch)
  File "/root/miniconda/lib/python3.7/site-packages/textattack/models/wrappers/huggingface_model_wrapper.py", line 24, in model_predict
    input_dict = {k: [_dict[k] for _dict in inputs] for k in inputs[0]}
  File "/root/miniconda/lib/python3.7/site-packages/textattack/models/wrappers/huggingface_model_wrapper.py", line 24, in <dictcomp>
    input_dict = {k: [_dict[k] for _dict in inputs] for k in inputs[0]}
  File "/root/miniconda/lib/python3.7/site-packages/textattack/models/wrappers/huggingface_model_wrapper.py", line 24, in <listcomp>
    input_dict = {k: [_dict[k] for _dict in inputs] for k in inputs[0]}
IndexError: list index out of range

```

**Comment by jxmorris12** [comment]
What are `model_path` and `tokenizer_path`? @zjiehang

**Comment by zjiehang** [comment]
```tokenizer_path``` is the path for "bert-base-uncased".
And ```model_path``` is the path for my finetune model,  based on "bert-base-uncased".

I find this bugs come from the tokenizer methods in HuggingFaceModelWrapper
In HuggingFaceModelWrapper, try self.tokenizer.encode_plus() for each text?

Another bug may happen, for the length of input_ids after tokenizing in HuggingFaceModelWrapper. We cann't make sure the length of input_ids always be the same, for BertTokenizer uses subword. Hope you can check it! Thanks!

**Comment by jxmorris12** [comment]
Fixed in v0.2.5. Here's the relevant Colab notebook: https://colab.research.google.com/drive/1amOmyG73Mc107sUzAlKzipkZp2M0x1nW?usp=sharing

**Comment by zjiehang** [comment]
Thanks for your works! Bug above is fixed.
However, something gets slower. For a batch whose max_length is 256 (defined by AutoTokenizer), we may not need 256 in fact. ?
For example, for a batch ['I am a cat', 'It is a dog'.], its actual max_length is 4-6 (considering <CLS> in Bert).
Maybe truncated the max_length in model_predict in HuggingFaceModelWrapper? Thanks!

**Comment by ziqi-zhang** [comment]
@jxmorris12 Hi I encountered the same problem and I didn't the v0.2.5 version of the code. I inspect the code and found that the inputs variable given to _model_predict function is a list of input_ids, instead of a dict.

**Comment by zjiehang** [comment]
@ziqi-zhang Bug is fixed in the newer version (v0.2.14), in which an AutoTokenizer is used. (See https://github.com/QData/TextAttack/blob/master/textattack/models/tokenizers/auto_tokenizer.py)
Try updating the textattack?

**Comment by ziqi-zhang** [comment]
@jxmorris12 Thanks for your quick reply! AutoTokenizer works. My problem was that I used the original hugginface's tokenizer and it needs to be wrapped by the AutoTokenizer of textattack.
